Modules

dsharletg opened the pull request:
This PR separates the gritty details of compilation from Func. It does this by encapsulating Halide IR in a few new types: Module and LoweredFunc. Func now has the ability to lower itself to a LoweredFunc (and Module), and then Modules can separately be compiled to LLVM (or C, or anything else). I of course kept the existing compile_to_* APIs for backwards compatibility, implemented in terms of the new output APIs.

Here are some motivations/advantages of this PR:
- It allows new lowering paths from Func without changing Halide itself. This could be useful for making plugins for things like Photoshop, Matlab, etc.
- This also easily enables some improvements like lowering multiple Funcs into one compiled object file.
- IMO, this also cleans up the CodeGen classes a bit. It makes the inputs/outputs simpler, and they have fewer lifetime issues that need to be managed.

abadams commented on src/CodeGen_LLVM.cpp:
```diff
@@ -2748,6 +2741,12 @@ void CodeGen_LLVM::visit(const Evaluate *op) {
 }
 
+void CodeGen_LLVM::visit(const Return *op) {
+    llvm::Value *value = codegen(op->value);
```
Shouldn't this follow a similar path to an assertion failure and clean up allocations (or assert that there are none)?

More generally, is adding a new IR node really necessary? This seems like it fights with the convention that Halide-generated functions (and Halide runtime functions) return exit codes.

dsharletg replied:
My original goal was to enable defining lowered functions that returned the value returned by another Halide function, in fact the motivation for adding Return was to preserve the return code of a Halide pipeline when called from another LoweredFunc: https://github.com/halide/Halide/pull/699/files#diff-c3cda5fa5b7d2d065eb79cb2c4d4ab2eR2091

In general, I think it will be pretty useful. One of the things easily enabled by this branch (and, maybe, the Return IR node) would be to make a LoweredFunc that dispatches to a bunch of other Halide pipelines, e.g. on GeneratorParams, or target architectures. I was picturing it with Return, but maybe this (and the above already existing example) works without Return? 

I think you're right about cleaning up/asserting there are no allocations. I will come back to that.

abadams commented on src/Module.h:
```diff
@@ -0,0 +1,84 @@
+
+namespace Internal {
+
+/** Definition of a lowered funciton. This object provides a concrete
```
http://esolangs.org/wiki/Funciton

abadams commented on src/CodeGen_LLVM.cpp:
```diff
@@ -426,63 +520,72 @@ void CodeGen_LLVM::compile(Stmt stmt, string name,
-    // Finally, verify the module is ok
-    internal_assert(verifyModule(*module) == false);
-    debug(2) << "Done generating llvm bitcode\n";
+void CodeGen_LLVM::visit(const Buffer *op) {
```
Seems weird to pass around a Buffer by const pointer. It looks like this is just for uniformity with IR nodes, but Buffer is not an IR Node. I suppose the same comment applies to LoweredFunc. It seems weird to pretend these things inherit from IRNode when they purposefully don't. Should the method even be called "visit"?

dsharletg replied:
Yeah, this could change. At first, I was trying to make it so Buffer (and even LoweredFunc) could exist as non-root IR nodes (e.g. declaring a Buffer as a local function object), and they were IR nodes when I started working on this. But, obviously that isn't something that works now. 

I will change these.

abadams commented on src/IRPrinter.cpp:
```diff
@@ -42,6 +42,21 @@ ostream &operator<<(ostream &stream, const Expr &ir) {
     return stream;
 }
 
+ostream &operator << (ostream &stream, const Buffer &buffer) {
```
operator<<(

steven-johnson commented on src/CodeGen_C.cpp:
```diff
@@ -181,7 +179,38 @@ const string preamble =
+    stream << "#ifdef __cplusplus\n";
+    stream << "#define HALIDE_EXTERN_C extern \"C\"\n";
+    stream << "#else\n";
+    stream << "#define HALIDE_EXTERN_C\n";
```
Wouldn't this be better as

    #define HALIDE_EXTERN_C extern\n

dsharletg replied:
I don't think this would work, the macro gets used even on function definitions. I think it might be an error to do:

    extern void f() { ... }

?

Either way, I don't think it will change anything. 'extern' is only necessary to get the behavior defined by extern "C".

zvookin replied:
Why are we using a macro at all instead of opening an extern C block before the definitions and closing it afterwards and protecting the block open and close with the standard C++ #ifdef conditional?

zvookin replied:
(Let me be elaborate a bit: for some reason people seem to really like the idea of using C macros to define their own programming language to decorate function declarations. I find this makes code much more difficult to read and it makes my life harder because the code might compile one way or the other and I have to think to figure out what is actually happening. It is sometimes necessary to do the macro thing because design of C/C++ is lacking in certain areas, but in this case I feel it can be avoided with no cost. The intent is the functions are always compiled with C linkage and the standard block scoping construct makes this clear.)

dsharletg replied:
It adds a small bit of complexity to do it this way (we can't put extern "C" around *everything* that CodeGen_C produces), I'll change it anyways though. I don't feel strongly about it. FWIW, this change is irrelevant of this PR, I just happened to notice that we were unconditionally generating `extern "C"` (i.e. compiling with a true C compiler would break) while changing this code around.

steven-johnson commented on src/CodeGen_LLVM.cpp:
```diff
@@ -302,12 +387,79 @@ void CodeGen_LLVM::compile(Stmt stmt, string name,
+namespace {
+
+// Make a wrapper to call the function with an array of pointer
+// args. This is easier for the JIT to call than a funtion with an
```
funtion, funciton, funyun, it's all good

dsharletg replied:
:)

dsharletg commented:
After speaking with @abadams offline, we decided to remove the Return IR node. I added it for two reasons: first, to preserve the return value of a call to another function (not just success/failure), and second, to avoid repeated error messages regarding the same failure.

We decided that the first issue is already decided by other bits of code, and that we would just add an (optional) parameter to AssertStmt if we need to preserve the return value. The second issue isn't necessarily that bad (it basically gives you a call stack of a failure). Andrew, correct me if I misunderstood :)

steven-johnson commented:
Any ETA on landing this?

abadams commented on src/Output.cpp:
```diff
@@ -0,0 +1,109 @@
+
+namespace Halide {
+
+void output_object(const Module &module, std::string filename) {
```
"output" is a very generic name. "compile_to" seems like a better prefix (ala the Func methods), but that seems weird for text or html. Any better ideas?

dsharletg replied:
I suppose they don't need to follow the same naming convention? We could use a mix of compile_to_*, export_*, write_*, etc.

Maybe a namespace for all of these functions is in order?

abadams commented on src/Output.cpp:
```diff
@@ -0,0 +1,109 @@
+
+    llvm::LLVMContext context;
+    llvm::Module *llvm = output_llvm_module(module, context);
+    output_object(llvm, object_filename);
```
I found it confusing that there's output_object(llvm::Module *, string) and also output_object(Halide::Module, string).

abadams commented:
It roughly LGTM, but it doesn't currently compile with trunk llvm, and it
needs testing on Windows too. Unfortunately, if it doesn't happen by
Wednesday morning then it's unlikely to happen until next week due to a
group ski trip.

On Mon, Mar 9, 2015 at 4:47 PM, Steven Johnson <notifications@github.com>
wrote:

> Any ETA on landing this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/699#issuecomment-77968890>.
>

dsharletg commented:
I figured I'd miss a few #ifdefs or something... I just updated my LLVM and kicked off the build, I'll take a look at it soon/tomorrow when its done.

dsharletg commented:
I also just remembered that a few of the tests are failing too. I'll hopefully be able to fix those up tomorrow  also.

dsharletg commented:
Well, I don't know if this still counts as "tomorrow" but I believe I've fixed the test issues I mentioned previously. It now at least builds on LLVM trunk, but there is a test failure on LLVM trunk that I just discovered (sigh).

dsharletg commented:
Alright, I believe I've fixed all known issues with this. I also renamed all the compile_halide_* to just compile_*.

I'm running a battery of tests right now, but in the last hour, its passed all of the tests, apps, OpenCL, CUDA, LLVM trunk, etc. I'm running them all again with the latest changes.

steven-johnson commented:
One more ping: how close are we to landing this?

dsharletg commented:
It's ready except for it just needs testing on Windows, which is running now.

steven-johnson commented:
One more weekly ping to see when this will land.

dsharletg commented:
I'm really sorry about the delay! We ran into problems on the Windows build machine (it's running out of space more and more often as the Halide build grows), so we've been bringing up a new Windows build machine. It's also much faster in general, so we should be able to test stuff on windows much more easily now.

dsharletg commented:
I've merged master, and I fixed a problem that revealed itself on windows. The windows tests are re-running now, if they pass I think this is ready to merge.

dsharletg commented:
Everything passed on Windows, finally! Sorry again about the delay.
